pkg/tinyvcdiff/contrib/tinyvcdiff_mtd/mtd.c

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright (C) 2022 Juergen Fitschen
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+#include "vcdiff_mtd.h"
+
+static int _erase (void *dev, size_t offset, size_t len)
+{
+    int rc;
+    vcdiff_mtd_t *mtd = dev;
+    size_t addr = offset + len;
+    size_t page = addr / mtd->dev->page_size;
+    size_t sector = page / mtd->dev->pages_per_sector;
+    size_t cnt = sector - mtd->next_erase_sector + 1;
+
+    /* early exit if all sectors are already in the erased state */
+    if (cnt == 0) {
+        return 0;
+    }
+
+    /* erase sectors */
+    rc = mtd_erase_sector(mtd->dev, mtd->next_erase_sector, cnt);
+    if (rc != 0) {
+        return rc;
+    }
+
+    mtd->next_erase_sector += cnt;
+
+    return 0;
+}
+
+static int _read (void *dev, uint8_t *dest, size_t offset, size_t len)
+{
+    vcdiff_mtd_t *mtd = dev;
+
+    /* mtd_read_page will take care of calculation the right page */
+    int rc = mtd_read_page(mtd->dev, dest, 0, offset, len);
+
+    return rc;
+}
+
+static int _write (void *dev, uint8_t *src, size_t offset, size_t len)
+{
+    vcdiff_mtd_t *mtd = dev;
+
+    /* mtd_read_page will take care of calculation the right page
+     * it's safe to call mtd_read_page_raw: the erase driver already
+     * prepared the sector for writing */
+    int rc = mtd_write_page_raw(mtd->dev, src, 0, offset, len);
```

Comment:
Are you sure this is right? The input `offset` is offset in bytes, while the offset parameter to `mtd_write_page_raw` is the offset in the page being written to, shouldn't it be something like:

```suggestion
    uint32_t page = offset / mtd->dev->page_size;
    uint32_t page_offset = (offset % mtd->dev->page_size);
    int rc = mtd_write_page_raw(mtd->dev, src, page, page_offset, len);
```

(untested)

Reply:
Same question goes for the read function actually.

Reply:
It should work to start at page `0` and let the MTD implemenation [figure out the right page](https://github.com/RIOT-OS/RIOT/blob/fca56ba0c3383f713179546e1a3a9b5bc2f24fd8/drivers/mtd/mtd.c#L255-L256).

Reply:
Well, but another problem could occur: un-aligned writes (i.e. `addr % FLASHPAGE_WRITE_BLOCK_ALIGNMENT` and `size % FLASHPAGE_WRITE_BLOCK_SIZE` are possible troublemaker ...). Or am I wrong?

Reply:
yea but you don't get those through MTD - #17683 tries to fix that

Reply:
I added write alignment for the MTD backend: b705803